cloud/aws/providerconfig/v1alpha1/types.go

```diff
@@ -234,3 +238,65 @@ func (s Subnets) FilterPublic() (res Subnets) {
 type RouteTable struct {
 	ID string `json:"id"`
 }
+
+// SecurityGroupRole defines the unique role of a security group.
+type SecurityGroupRole string
+
+var (
+	SecurityGroupNode         = SecurityGroupRole("node")
+	SecurityGroupControlPlane = SecurityGroupRole("controlplane")
+)
+
+// SecurityGroup defines an AWS security group.
+type SecurityGroup struct {
+	ID   string `json:"id"`
+	Name string `json:"name"`
+
+	IngressRules IngressRules `json:"ingressRule"`
+}
+
+// String returns a string representation of the security group.
+func (s *SecurityGroup) String() string {
+	return fmt.Sprintf("id=%s/name=%s", s.ID, s.Name)
+}
+
+// IngressRule defines an AWS ingress rule for security groups.
+type IngressRule struct {
+	Description string `json:"description"`
+	Protocol    string `json:"protocol"`
+	FromPort    int64  `json:"fromPort"`
+	ToPort      int64  `json:"toPort"`
+
+	// List of CIDR blocks to allow access from. Cannot be specified with SourceSecurityGroupID.
+	CidrBlocks []string `json:"cidrBlocks"`
+
+	// The security group id to allow access from. Cannot be specified with CidrBlocks.
+	SourceSecurityGroupID *string `json:"sourceSecurityGroupId"`
+}
+
+// String returns a string representation of the ingress rule.
+func (i *IngressRule) String() string {
+	return fmt.Sprintf("protocol=%s/range=[%d-%d]/description=%s", i.Protocol, i.FromPort, i.ToPort, i.Description)
+}
+
+// IngressRules is a slice of AWS ingress rules for security groups.
+type IngressRules []*IngressRule
+
+// Returns the difference between this slice and the other slice.
+func (i IngressRules) Difference(o IngressRules) (out IngressRules) {
+	for _, x := range i {
+		found := false
+		for _, y := range o {
+			if reflect.DeepEqual(x, y) {
```

Comment:
i would strongly prefer to define equality on an `IngressRule` and use that here instead of deep equal, but will be ok punting if you open an issue and make a comment that this should be replaced.